### ISSUE TITLE
Declaration for controlling {left/right} association

### ISSUE THREAD

**Issue by ehiggs** [kevinmehall/rust-peg#148]
PEG is known to associate to the right when there is precedence climbing. However, some operators should be left associative; e.g. when repeated.

e.g.
```
expression -> i64
	= sum

sum -> i64
	= l:product "+" r:sum { l+r }
	/ l:product "-" r:sum { l-r }
	/ product

product -> i64
	= l:pow "*" r:product { l*r }
	/ l:pow "/" r:product { l/r }
	/ pow 

pow -> i64
    = l:atom "^" r:pow { l.pow(r as u32) }
    / atom

atom -> i64
	= number
	/ "(" v:sum ")" { v }

number -> i64
	= [0-9]+ { match_str.parse().unwrap() }
```

```
	assert_eq!(expression("2^2^2^2"), Ok(65536)); // <-- pow is usually right assoc. e.g. python
	assert_eq!(expression("1024/2/2/2"), Ok(128)); // <-- fails because it should be left associative.
```

I propose that there should be a declaration like `#[left_assoc]` or `#[assoc=left]` which can be applied to a rule which will change the parsing behaviour as appropriate.

```
#[left_assoc]
pow -> i64
    = l:atom "^" r:pow { l.pow(r as u32) }
    / atom
```

**Comment by kevinmehall** [comment]
Could you elaborate on what code you'd expect this declaration to generate? If you specify an `atom` rule on the left and the recursive rule on the right, the resulting tree will be right associative. `l:pow "^" r:atom` would be how you'd represent a left-associative tree, but left recursive rules lead to infinite recursion in most PEG implementations. There are [ways to fix that](http://tratt.net/laurie/research/pubs/html/tratt__direct_left_recursive_parsing_expression_grammars/), though.

The way PEG grammars typically handle associativity is by using `*` repetition instead of recursion, and then using action code to produce a tree if desired from the resulting list of tokens. I've been thinking about how rust-peg could make this easier and I'm considering integrating a [precedence climbing](http://eli.thegreenplace.net/2012/08/02/parsing-expressions-by-precedence-climbing) parser, such that you could specify the associativity and precedence of a set of operators (syntax TBD):

```
expression -> i64
   = #infix(atom) |l, r| {
       #L "+" { l + r }
          "-" { l - r }
       #L "*" { l * r }
          "/" { l / r }
       #R "^" { l.pow(r) }
   }
```

which would essentially parse `(atom (("*" / "+" / "-" / "/" / "^") atom)*` using traditional PEG semantics, while then running a precedence-climbing parser over the resulting tokens instead of building a Vec.

**Comment by kevinmehall** [comment]
I've prototyped precedence-climbing on the [infix branch](https://github.com/kevinmehall/rust-peg/compare/infix). I think the syntax could be better, so I'm going to experiment further before merging, but let me know what you think.

**Comment by ehiggs** [comment]
>Could you elaborate on what code you'd expect this declaration to generate? If you specify an atom rule on the left and the recursive rule on the right, the resulting tree will be right associative. l:pow "^" r:atom would be how you'd represent a left-associative tree, but left recursive rules lead to infinite recursion in most PEG implementations. 

That's exactly the issue. We can fix the parse tree post parse to pivot repeated operators so their associativity is corrected. But this a hack and there are other peg parsers that solve this at the parse level.

I hadn't considered using `*` and constructing the tree explicitly. That's not a bad idea, tbh.

The proposed syntax looks ok, tbh. [Pest](https://github.com/dragostis/pest) uses `<` as in the following `exponention` rule:

```rust
    grammar! {
        // precedence climbing
        expression = _{ // rule is silent because it's the rule we're matching
            { ["("] ~ expression ~ [")"] | number } // primary
            addition       = { plus  | minus } // precedence 0
            multiplication = { times | slash } // precedence 1
            exponention    = {< pow }          // precedence 2; NB: exponention is right associative.
        }
        number = @{ ["-"]? ~ (["0"] | ['1'..'9'] ~ ['0'..'9']*) } // atomic because it cannot
        plus   =  { ["+"] }                                       // accept white-space
        minus  =  { ["-"] }
        times  =  { ["*"] }
        slash  =  { ["/"] }
        pow    =  { ["^"] } 

        whitespace = _{ [" "] } // whitespce gets run between all rules
    }
```

I think `#L` and `#R` are short and almost descriptive enough to get a sense of the meaning while still being notational. I think Pest's operator is a bit too notational since it's unsearchable. 

Regarding your infix operator, do they fall through or default to the right? You have `+` decorated but not `-`.

**Comment by kevinmehall** [comment]
Released 0.5 with a tweaked version of the above proposal. [Docs here](https://github.com/kevinmehall/rust-peg#infix-expressions). Let me know if you think of ways to improve the functionality or the documentation.
